### ISSUE TITLE
NiftiMapsMasker takes a large amount of time to perform .transform when using DiFuMo 1024 atlas

### ISSUE THREAD

**Issue by RaphaelMeudec** [nilearn/nilearn#3177]
Nilearn version: 0.8.1

### Expected behavior

Computation is quite long for DiFuMo 1024 on a simple image. I guess there might be some optimizations that could be done to quicken the computation?

### Steps and code to reproduce bug

```python
from nilearn import datasets, input_data

difumo = datasets.fetch_atlas_difumo(dimension=1024, resolution_mm=3)
difumo_masker = input_data.NiftiMapsMasker(maps_img=difumo.maps, verbose=0).fit()

sample_image = datasets.fetch_neurovault_motor_task().images[0]
niftimapsmasker_difumo = difumo_masker.transform(sample_image)
```

**Comment by bthirion** [comment]
Takes 14s on a big server

**Comment by htwangtw** [comment]
We observe the same run time here in my lab. It's about 20 minutes when I don't limit memory usage but limit the task on one core on our server. I had a lot of trail and error to optimise the resource needed when running any probability atlases.
I have learned a few things to slightly improve the speed, as for now:

1. resample your atlas to the target shape (data by default) - this takes up a lot of run time even for `LabelMasker`. 
2. Use `mask_img` to reduce the data point fitted during time series extraction. fMRIPrep produces a mask per functional image. In the current implementation, `mask_img` is supplied by the user when creating a masker object. 

I was working on a bench mark for this a while ago. Was waiting for my schedule to be less busy to further investigate this issue and do some experiments.

cc @ltetrel who might have some ideas as well

**Comment by RaphaelMeudec** [comment]
Iterating on @htwangtw observations, here are some ideas that could improve the NiftiMapsMasker IMO:

- the cache mechanism applies at multiple level: on the overall computations, but also on the resampling of the maps. This means that for different input images but with same shape/affine, the cache applies partially to resample the maps. This is important as the resampling is done in the `.transform` and it's not clear in the dosctring.
  ```
  memory : joblib.Memory or str, optional
        Used to cache the region extraction process.
        By default, no caching is done. If a string is given, it is the
        path to the caching directory.
  ```
- a parallel call for the resampling of the maps which would quicken the computations to [replace this block](https://github.com/nilearn/nilearn/blob/1ac949da92b06fb028fe2d94e29d089eb0ba98a7/nilearn/maskers/nifti_maps_masker.py#L460):
  ```python
            if not _check_same_fov(ref_img, self._resampled_maps_img_):
                if self.verbose > 0:
                    print("Resampling maps")

                def parallel_resample_img(maps_imgs_, **kwargs):
                    """Example of parallel implementation that could be used"""
                    import joblib

                    n_jobs = joblib.cpu_count()
                    results = joblib.Parallel(n_jobs=n_jobs)(
                        joblib.delayed(image.resample_img)(
                            map_img_,
                            **kwargs
                        ) for map_img_ in image.iter_img(maps_imgs_)
                    )

                    return image.concat_imgs(results)

                self._resampled_maps_img_ = parallel_resample_img(
                    self.maps_img_, interpolation="continuous",
                    target_shape=ref_img.shape[:3],
                    target_affine=ref_img.affine)
  ```

@jeromedockes maybe you'll have some insights on this?

**Comment by jeromedockes** [comment]
![profile](https://user-images.githubusercontent.com/9196501/162479282-368ec7a0-1f47-4bfb-9bc0-99742d782df9.svg)
here is a profiling of the script provided by @RaphaelMeudec ; not surprisingly most of the time is spent resampling and regressing the inputs on the atlas maps

**Comment by bthirion** [comment]
To reduce computation time, we could learn the pseudo inverse of the masks matrix during fit, and simply do a dot product at transform time.

**Comment by bthirion** [comment]
Which implies that the images would be resamples to the atlas during transform, and not the converse.

**Comment by Remi-Gau** [comment]
Note that if this script is used now, you will get a warning:


```
[fetch_atlas_difumo] Dataset found in /home/remi-gau/nilearn_data/difumo_atlases
[NiftiMapsMasker.fit] loading regions from /home/remi-gau/nilearn_data/difumo_atlases/1024/3mm/maps.nii.gz
/home/remi-gau/github/nilearn/nilearn/.venv/lib/python3.13/site-packages/sklearn/utils/_set_output.py:316: UserWarning: Resampling maps at transform time...
To avoid this warning, make sure to pass the images you want to transform to fit() first, or directly use fit_transform().
  data_to_wrap = f(self, X, *args, **kwargs)
[NiftiMapsMasker.wrapped] Loading data from /home/remi-gau/github/nilearn/nilearn/nilearn/datasets/data/image_10426.nii.gz
[NiftiMapsMasker.wrapped] Extracting region signals
[NiftiMapsMasker.wrapped] Cleaning extracted signals
```

So if you are planning to extract data from several images, better to pass them to fit() first, otherwise you will be resampling the maps with every image to transform.

```python
from nilearn import datasets, maskers

difumo = datasets.fetch_atlas_difumo(dimension=1024, resolution_mm=3)

sample_image = datasets.load_sample_motor_activation_image()

difumo_masker = maskers.NiftiMapsMasker(maps_img=difumo.maps, verbose=1).fit(sample_image)
niftimapsmasker_difumo = difumo_masker.transform(sample_image)
```

**Comment by bthirion** [comment]
Is this something we want to make explicit in the documentation ?

**Comment by htwangtw** [comment]
I agree. 

It might be worth it to explain what `fit()` does in the context of nilearn. I can see from user's perceptive, if they just want to transform their data and the majority of the `fit_transform(img)` operation doesn't take that much time, then it's logical to default to running `fit_transform(img)` directly.
